File: relay/liveactivity.go
package relay

import (
	"encoding/json"
	"regexp"
	"sort"
	"time"
)

// Live Activities (native/spec/push.md §7). ActivityKit gives the app one
// push token per activity (its updates) and one for the app (push-to-start).
// The app registers each as a Live Activity handle hanging off the device
// handle it made for the same workspace (POST /v1/handles {pushType:
// "liveactivity", parent}); the workspace then pushes content states to it
// (POST /v1/push {type: "liveactivity", activity}).
//
// An ActivityKit payload can't be sealed: the system decodes the content
// state and draws it before any code of the app's runs. So the relay does
// not forward a payload at all — it builds the APNs body itself from a few
// enumerated and numeric fields, and the only strings that pass are two
// opaque ids of a push-to-start. No title, text or path reaches APNs this
// way; what a workspace can say through it is "running / waiting / idle,
// since, N pending".

// PushTypeLiveActivity is the handle type and push type of Live Activities.
const PushTypeLiveActivity = "liveactivity"

// ActivityAttributesType is the app's ActivityAttributes type — the
// "attributes-type" of a push-to-start (native/ios: XbinAgent's
// AgentActivityAttributes).
const ActivityAttributesType = "AgentActivityAttributes"

const (
	// maxChildren is how many Live Activity handles one device handle
	// holds; a new one evicts the least recently used.
	maxChildren = 16
	maxPending  = 99
	// liveTopicSuffix makes the APNs topic of a Live Activity push.
	liveTopicSuffix = ".push-type.liveactivity"
)

var liveRefRe = regexp.MustCompile(`^[A-Za-z0-9_-]{1,64}$`)

// ActivityState is a Live Activity's content state — the app's
// AgentActivityState, key for key.
type ActivityState struct {
	Phase   string `json:"phase"`   // running | waiting | idle
	Since   int64  `json:"since"`   // unix seconds the turn started (0 = unknown)
	Pending int    `json:"pending"` // requests waiting for the user (0–99)
}

// ActivityPush is the "activity" of a liveactivity push.
type ActivityPush struct {
	Event         string         `json:"event"`     // update | end | start
	Timestamp     int64          `json:"timestamp"` // unix seconds; the device ignores anything older than what it shows
	State         *ActivityState `json:"state"`
	StaleDate     int64          `json:"staleDate,omitempty"`     // unix seconds
	DismissalDate int64          `json:"dismissalDate,omitempty"` // end only
	// start only: opaque ids the app maps back (xbind's push id for the
	// workspace, and the reference the app registers the new activity's
	// token under)
	WS  string `json:"ws,omitempty"`
	Ref string `json:"ref,omitempty"`
}

// check validates a with the relay's clock; "" = fine.
func (a *ActivityPush) check(now time.Time) string {
	soon := now.Add(time.Hour).Unix()
	later := now.Add(24 * time.Hour).Unix()
	switch {
	case a.Event != "update" && a.Event != "end" && a.Event != "start":
		return "activity.event: update | end | start"
	case a.Timestamp <= 0 || a.Timestamp > soon:
		return "activity.timestamp: unix seconds, not in the future"
	case a.State == nil:
		return "activity.state: {phase, since, pending}"
	case a.State.Phase != "running" && a.State.Phase != "waiting" && a.State.Phase != "idle":
		return "activity.state.phase: running | waiting | idle"
	case a.State.Since < 0 || a.State.Since > soon:
		return "activity.state.since: unix seconds"
	case a.State.Pending < 0 || a.State.Pending > maxPending:
		return "activity.state.pending: 0–99"
	case a.StaleDate < 0 || a.StaleDate > later:
		return "activity.staleDate: unix seconds, within a day"
	case a.DismissalDate < 0 || a.DismissalDate > later || (a.DismissalDate != 0 && a.Event != "end"):
		return "activity.dismissalDate: unix seconds within a day, end only"
	}
	if a.Event == "start" {
		if !liveRefRe.MatchString(a.WS) || !liveRefRe.MatchString(a.Ref) {
			return "activity.ws, activity.ref: 1–64 of A–Z a–z 0–9 _ - (start)"
		}
	} else if a.WS != "" || a.Ref != "" {
		return "activity.ws, activity.ref: start only"
	}
	return ""
}

// startAlert is the generic alert of a push-to-start (ActivityKit shows one
// when an activity starts by push), by phase.
func startAlert(phase string) map[string]string {
	body := "An agent is working."
	switch phase {
	case "waiting":
		body = "An agent is waiting for you."
	case "idle":
		body = "An agent finished."
	}
	return map[string]string{"title": "xbin", "body": body}
}

// activityBody builds the APNs body of a liveactivity push from checked
// fields only.
func activityBody(a *ActivityPush) ([]byte, error) {
	aps := map[string]any{
		"timestamp":     a.Timestamp,
		"event":         a.Event,
		"content-state": ActivityState{Phase: a.State.Phase, Since: a.State.Since, Pending: a.State.Pending},
	}
	if a.StaleDate != 0 {
		aps["stale-date"] = a.StaleDate
	}
	if a.DismissalDate != 0 {
		aps["dismissal-date"] = a.DismissalDate
	}
	if a.Event == "start" {
		aps["attributes-type"] = ActivityAttributesType
		// the display names are the app's to fill in (from its own records
		// of ws); a push never carries them
		aps["attributes"] = map[string]string{"ws": a.WS, "ref": a.Ref, "workspace": "", "session": "",
			"appWorkspace": "", "sessionID": ""}
		aps["alert"] = startAlert(a.State.Phase)
		// ask for the new activity's update token (the app registers it)
		aps["input-push-token"] = 1
	}
	return json.Marshal(map[string]any{"aps": aps})
}

// newChild stores a Live Activity handle under the device handle parent.
// The same token under the same parent is the same handle (the app
// re-registers after a restart); a parent holds at most maxChildren.
func (s *store) newChild(parent, token, topic, env string, now time.Time) (string, error) {
	s.mu.Lock()
	recs := s.sweepLocked(now, false)
	p := s.st.Handles[parent]
	switch {
	case p == nil:
		s.mu.Unlock()
		_ = s.write(recs)
		return "", errNoHandle
	case p.Type != "" || p.Topic != topic || p.Env != env:
		s.mu.Unlock()
		_ = s.write(recs)
		return "", errHandleType
	}
	for c := range s.children[parent] {
		if h := s.st.Handles[c]; h != nil && h.Token == token {
			s.mu.Unlock()
			return c, s.write(recs)
		}
	}
	if len(s.st.Handles) >= s.lim.maxHandles {
		recs = append(recs, s.sweepLocked(now, true)...)
		if len(s.st.Handles) >= s.lim.maxHandles || s.st.Handles[parent] == nil {
			s.mu.Unlock()
			_ = s.write(recs)
			return "", errFull
		}
	}
	if kids := s.children[parent]; len(kids) >= maxChildren {
		ids := make([]string, 0, len(kids))
		for c := range kids {
			ids = append(ids, c)
		}
		sort.Slice(ids, func(i, j int) bool {
			a, b := s.st.Handles[ids[i]], s.st.Handles[ids[j]]
			return max(a.Used, a.Created) < max(b.Used, b.Created)
		})
		for _, c := range ids[:len(ids)-maxChildren+1] {
			recs = append(recs, s.delTree(c)...)
		}
	}
	id := randomID(16)
	s.st.Handles[id] = &Handle{Token: token, Topic: topic, Env: env, Type: PushTypeLiveActivity, Parent: parent, Created: now.Unix()}
	s.link(parent, id)
	recs = append(recs, s.putH(id))
	s.mu.Unlock()
	return id, s.write(recs)
}

// handleType reports a handle's type ("" or "liveactivity").
func (s *store) handleType(id string) (string, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if h := s.st.Handles[id]; h != nil {
		return h.Type, true
	}
	return "", false
}

File: native/ios/Packages/XbinAgent/Sources/XbinAgent/LiveActivity.swift
import Foundation

// Live Activities for agent turns (native/spec/push.md §7): the lock-screen
// and Dynamic Island card of a turn that runs. Everything that doesn't draw
// is here — the content state and attributes (their JSON is the relay's
// wire), what a transcript says the card shows, when the app starts,
// updates and ends one, and the strings the widget draws — so it is tested
// on Linux. The ActivityKit conformance, the widget and the app's glue are
// in native/ios/Widgets and App/Push.
//
// The card is generic on purpose: an ActivityKit payload can't be sealed,
// so what xbind pushes through the relay is only the phase, since when and
// how many requests wait. Names (the workspace's, the session's) exist only
// on the device: the app puts them in the attributes when it starts an
// activity itself; one started by push carries none, and the widget looks
// the workspace up by `ws` in the app's own records.

/// A turn as the card shows it.
public enum AgentActivityPhase: String, Sendable, Hashable, Codable, CaseIterable {
    /// The agent is working.
    case running
    /// A permission request or a question waits for the user.
    case waiting
    /// The turn is over (the card's last state before it goes).
    case idle
}

/// The content state (ActivityKit's `ContentState`). Its JSON is what the
/// relay builds (`{"phase", "since", "pending"}`); decoding is lenient — an
/// unknown phase reads as running, a missing number as 0 — so a newer
/// xbind never leaves the card undecodable.
public struct AgentActivityState: Sendable, Hashable, Codable {
    public var phase: AgentActivityPhase
    /// Unix seconds the turn started; 0 = unknown (no elapsed time shown).
    public var since: Int64
    /// Permission requests and questions waiting.
    public var pending: Int

    public init(phase: AgentActivityPhase, since: Int64 = 0, pending: Int = 0) {
        self.phase = phase
        self.since = since
        self.pending = pending
    }

    enum CodingKeys: String, CodingKey { case phase, since, pending }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let raw = (try? c.decodeIfPresent(String.self, forKey: .phase)) ?? nil
        phase = raw.flatMap(AgentActivityPhase.init(rawValue:)) ?? .running
        since = Self.int64(c, .since)
        pending = Int(clamping: Self.int64(c, .pending))
    }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(phase, forKey: .phase)
        try c.encode(since, forKey: .since)
        try c.encode(pending, forKey: .pending)
    }

    /// A whole number, whether the coder wrote it as an integer or a double.
    private static func int64(_ c: KeyedDecodingContainer<CodingKeys>, _ k: CodingKeys) -> Int64 {
        if let v = try? c.decodeIfPresent(Int64.self, forKey: k) { return v }
        if let d = try? c.decodeIfPresent(Double.self, forKey: k), d.isFinite { return Int64(d) }
        return 0
    }

    /// When the turn started, when known.
    public var startDate: Date? { since > 0 ? Date(timeIntervalSince1970: TimeInterval(since)) : nil }

    /// The card's last state for a turn that ended.
    public var finished: AgentActivityState { AgentActivityState(phase: .idle, since: since, pending: 0) }
}

/// The attributes (ActivityKit's `ActivityAttributes`, fixed for an
/// activity's life). The conformance itself is declared where ActivityKit
/// is (native/ios/Widgets/Shared); a push-to-start names this type
/// ("attributes-type": "AgentActivityAttributes") and fills only `ws` and
/// `ref`. Decoding is lenient: a missing field reads as "".
public struct AgentActivityAttributes: Sendable, Hashable, Codable {
    public typealias ContentState = AgentActivityState

    /// xbind's push id of the workspace (`workspace` of the push
    /// registration; "" when the app has none yet).
    public var ws: String
    /// The reference of an activity xbind started by push: the app
    /// registers the activity's token under it ("" for the app's own).
    public var ref: String
    /// Display names, set by the app ("" in a push-started one: the widget
    /// looks the workspace up by `ws`).
    public var workspace: String
    public var session: String
    /// The app's workspace id and the agent session id (the card's link);
    /// "" in a push-started one.
    public var appWorkspace: String
    public var sessionID: String

    public init(ws: String = "", ref: String = "", workspace: String = "", session: String = "",
                appWorkspace: String = "", sessionID: String = "") {
        self.ws = ws
        self.ref = ref
        self.workspace = workspace
        self.session = session
        self.appWorkspace = appWorkspace
        self.sessionID = sessionID
    }

    enum CodingKeys: String, CodingKey { case ws, ref, workspace, session, appWorkspace, sessionID }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        func s(_ k: CodingKeys) -> String { ((try? c.decodeIfPresent(String.self, forKey: k)) ?? nil) ?? "" }
        self.init(ws: s(.ws), ref: s(.ref), workspace: s(.workspace), session: s(.session),
                  appWorkspace: s(.appWorkspace), sessionID: s(.sessionID))
    }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(ws, forKey: .ws)
        try c.encode(ref, forKey: .ref)
        try c.encode(workspace, forKey: .workspace)
        try c.encode(session, forKey: .session)
        try c.encode(appWorkspace, forKey: .appWorkspace)
        try c.encode(sessionID, forKey: .sessionID)
    }

    /// Started by push (xbind), not by the app.
    public var pushStarted: Bool { !ref.isEmpty }
}

extension AgentTranscript {
    /// What the card shows now: nil between turns (from the turn's end on,
    /// even while the idle status is still to come, and once the session is
    /// over). Waiting when the agent says so or a request is unanswered;
    /// `since` is the turn's start (xbind computes the same from the log).
    public var activityState: AgentActivityState? {
        guard state.isBusy, let started = state.turnStartedAt else { return nil }
        let n = min(pendingPermissions.count + pendingQuestions.count, 99)
        let phase: AgentActivityPhase = (state.status == .waitingPermission || n > 0) ? .waiting : .running
        return AgentActivityState(phase: phase, since: started / 1000, pending: n)
    }
}

/// When the app starts, updates and ends a card (the app decides locally
/// while it runs; xbind pushes the rest). A card appears for a turn that
/// has run `startAfter` (quick turns at the desk never flash one), at once
/// for one that waits for the user, or when the app is about to leave the
/// foreground mid-turn (ActivityKit starts activities only from the
/// foreground); it follows the state; it ends with the turn, staying
/// `dismissAfter` on the lock screen.
public struct AgentActivityPolicy: Sendable, Hashable {
    public var startAfter: TimeInterval
    public var dismissAfter: TimeInterval
    /// How long a running card may go without news before it reads stale.
    public var staleAfter: TimeInterval

    public init(startAfter: TimeInterval = 10, dismissAfter: TimeInterval = 600, staleAfter: TimeInterval = 4 * 3600) {
        self.startAfter = startAfter
        self.dismissAfter = dismissAfter
        self.staleAfter = staleAfter
    }

    public enum Action: Sendable, Hashable {
        case none
        case start(AgentActivityState)
        case update(AgentActivityState)
        /// The final state, and when the system should remove the card.
        case end(AgentActivityState, dismissAt: Date)
    }

    public struct Decision: Sendable, Hashable {
        public var action: Action
        /// Decide again then (a turn that is not old enough yet).
        public var recheckAt: Date?

        public init(_ action: Action, recheckAt: Date? = nil) {
            self.action = action
            self.recheckAt = recheckAt
        }
    }

    /// - Parameters:
    ///   - state: the session's `activityState` (nil between turns).
    ///   - shown: what the card shows (nil = no card).
    ///   - canStart: the app is in the foreground and may start one (the
    ///     user allows Live Activities and did not dismiss this turn's).
    ///   - leaving: the app is about to leave the foreground.
    public func decide(state: AgentActivityState?, shown: AgentActivityState?, canStart: Bool,
                       leaving: Bool = false, now: Date) -> Decision {
        switch (state, shown) {
        case (nil, nil):
            return Decision(.none)
        case (nil, let s?):
            return Decision(.end(s.finished, dismissAt: now.addingTimeInterval(dismissAfter)))
        case (let s?, let cur?):
            return Decision(s == cur ? .none : .update(s))
        case (let s?, nil):
            guard canStart else { return Decision(.none) }
            if s.phase == .waiting || leaving { return Decision(.start(s)) }
            guard let start = s.startDate else {
                return Decision(.none) // no clock to count from: only a wait or leaving starts it
            }
            let due = start.addingTimeInterval(startAfter)
            return due <= now ? Decision(.start(s)) : Decision(.none, recheckAt: due)
        }
    }
}

/// What the card says, as strings and symbol names — the widget only lays
/// them out.
public struct AgentActivityDisplay: Sendable, Hashable {
    /// The session's name, else "Agent".
    public var title: String
    /// The workspace's name, else "xbin".
    public var subtitle: String
    /// "Working", "Waiting for you", "2 waiting for you", "Finished";
    /// "Status unknown" when stale and not finished.
    public var status: String
    /// A short form for the compact Dynamic Island ("", "!", "2").
    public var badge: String
    /// SF Symbol.
    public var symbol: String
    public var phase: AgentActivityPhase
    public var stale: Bool
    /// Count elapsed time from here (nil: unknown, or finished).
    public var timerStart: Date?
    /// `xbin://<app workspace>/agent/<session>`, `xbin://<app workspace>`,
    /// or nil (a card the app can't place).
    public var link: String?

    /// - Parameters:
    ///   - workspaceTitle, appWorkspace: the app's own records of
    ///     `attributes.ws` (for a push-started card, which carries neither).
    public init(attributes a: AgentActivityAttributes, state s: AgentActivityState, stale: Bool,
                workspaceTitle: String? = nil, appWorkspace: String? = nil) {
        title = a.session.isEmpty ? "Agent" : a.session
        let ws = a.workspace.isEmpty ? (workspaceTitle ?? "") : a.workspace
        subtitle = ws.isEmpty ? "xbin" : ws
        phase = s.phase
        self.stale = stale && s.phase != .idle
        switch s.phase {
        case .running:
            status = "Working"
            symbol = "sparkles"
            badge = ""
        case .waiting:
            status = s.pending > 1 ? "\(s.pending) waiting for you" : "Waiting for you"
            symbol = "hand.raised.fill"
            badge = s.pending > 1 ? "\(s.pending)" : "!"
        case .idle:
            status = "Finished"
            symbol = "checkmark.circle.fill"
            badge = ""
        }
        if self.stale {
            status = "Status unknown"
            symbol = "questionmark.circle"
        }
        timerStart = s.phase == .idle || self.stale ? nil : s.startDate
        let app = a.appWorkspace.isEmpty ? (appWorkspace ?? "") : a.appWorkspace
        link = Self.link(appWorkspace: app, session: a.sessionID)
    }

    /// The card's deep link (plans/native.md §4: `xbin://<ws>/agent/<id>`).
    static func link(appWorkspace: String, session: String) -> String? {
        let allowed = CharacterSet.urlHostAllowed.subtracting(CharacterSet(charactersIn: "/?#@:"))
        guard !appWorkspace.isEmpty, let ws = appWorkspace.addingPercentEncoding(withAllowedCharacters: allowed) else { return nil }
        guard !session.isEmpty, let id = session.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed.subtracting(CharacterSet(charactersIn: "/?#")))
        else { return "xbin://\(ws)" }
        return "xbin://\(ws)/agent/\(id)"
    }
}
